proposals/lockout-violation-detection.md

```diff
@@ -1,56 +1,96 @@
+If there are any votes made that have a root `R` missing from the rooted
+SlotHashes list, slash them on the rooted fork since they have committed
+to a different fork
+
+#### Removing lockouts
+Attempting to illegally *remove* a lockout in a newer vote that that should
+have still existed based on an older vote is a lockout violation. We detect
+this as follows:
+
+1. For all other slots `S` in each vote other than the root, track a range
 `(S, S + 2^n)` where `n` is the confirmation count. For each slot `S` we only
 have to track the greatest such lockout range in the database.
-4) For each new vote `V`, for each slot `S` in the vote, lookup in the database
+
+2. For each new vote `V`, for each slot `S` in the vote, lookup in the database
 to see if there's some range `(X, X + 2^n)` where `S` is in this range, but
-`X < S` and `X` is missing from the vote `V`. This is a lockout violation
-because this implies that the validator made a vote where `S` popped off `X`,
-but the lockout on `X` from an earlier vote should have prevented that from
-happening.
+`X` is missing from the vote `V`. This is a lockout violation because this
+implies that the validator made a vote where `S` popped off `X`, but the
+lockout on `X` from an earlier vote should have prevented that from happening.
 
 Note for each interval `(S, S + 2^n)` we also need to be able to lookup the
 vote. This is important for being able to pull up the votes later as part of a
 proof.
 
-5) For each validator we also track a rooted set in the database as well to
-catch violating lockouts on rooted slots. We can remove an interval
-`(S, S + 2^n)` from the database once the slot becomes a root add it to a
-rooted set for this validator, and any new votes < root also get added to
+#### Reducing lockout
+Attempting to illegally *reduce* a lockout from an older vote in a newer vote
+is a lockout violation.
+
+The tricky bit of this is determining that one vote `V` is *older* than
+another vote `V'`. To this end, we track a flag `is_vote_earlier` whose usage
+we will highlight in the protocol below.
+
+1. We set `is_vote_earlier = false`.
+2. For each newly seen vote `V` made by a validator `X`, for each slot `S` in
+`V`:
+    - If `S` exists in the database:
+        - Compare the lockout `L_V` in the vote `V` on `S` against the greatest
+        lockout `L_D` in the database for that slot `S` made by that validator `X`.
+            - If `L_V` < `L_D`, set `is_vote_earlier=true`. 
+            - If `L_V` == `L_D`, continue.
+            - If `L_V > L_D`, check if the flag `is_vote_earlier=true`. If so,
+            this implies this vote `V` was *older* than some vote `V'` that
+            comitted the greater `S'` lockout to the database, yet `V` has a
+            lesser lockout on `S`, which means the validator reduced lockout on
+            `S` in a later vote. This is a lockout violation.
+    - If `S` does not exist in the database, the above `Removing lockouts`
```

Comment:
I don't think you need the If condition here this section describes reducing lockout violations which is only possible if `S` is in the database.

Reply:
I think this was just for clarity because there was a "If `S` exists in the database:" case above. Also for us it's clear that reducing lockout === must exist in database, but for first time readers it might not be so obvioius